stdlib/result.mli

```diff
@@ -0,0 +1,89 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*   Copyright 2018 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Result values.
+
+    @since 4.08 *)
+
+(** {1:results Results} *)
+
+type ('a, 'e) t = ('a, 'e) result = Ok of 'a | Error of 'e
+(** The type for result values. Either a value [Ok v] or an error [Error e]. *)
+
+val ok : 'a -> ('a, 'e) result
+(** [ok v] is [Ok v]. *)
+
+val error : 'e -> ('a, 'e) result
+(** [error e] is [Error e]. *)
```

Comment:
I also find this documentation string confusing. Similarly, maybe:

`[error v] returns the value [Error v] for any v.`

Reply:
(BTW, I recognize "is" is shorter, but "returns" feels more idiomatic to me in documentation.)

Reply:
The fact that I'm calling a function here is pointless hence no "return" needed. 

The doc string describes the resulting *value* and you can substitute the expression on the left of the `is` by what is mentioned on the right (and vice-versa) and your program will behave the same. 

That's the power of functional programming (something that somehow seems to have been lost in the FP hype). 

In other words I'm describing an equality here and `is` is the right word for this.

Reply:
You don't need documentation, you wrote the code. However, I found this "equality" opaque when I started reading the .mli file. The code is much more comprehensible than the docstring, and that's bad.

Reply:
> You don't need documentation, you wrote the code.

Don't know where you got this from I constantly need documentation for all the code I write.

This is not related to me writing the code or not,  it's about distilling a mindset on how to think about your programs: in terms of values, their denotation and their equalities rather than the mechanics of their evaluation. 

The doc string you propose is cumbersome and pure noise to me, it doesn't give me the essence of what the application stands for in code. With a little bit of practice I'm sure you'll eventually understand, get used to it and even grow fond of it.

Reply:
> The doc string you propose is cumbersome and pure noise to me

Replacing "is" with "returns" (which is far more idiomatic to most computer scientists I know) is cumbersome and pure noise? I can see you arguing that it is unnecessary, but _cumbersome_? _noisy_? That's a bit of a stretch, no?

The test here isn't whether _you_ think the docs are okay. It's whether others find them confusing or not. You can't tell me "you didn't find it confusing", I did. You can argue that I'm an imbecile, and perhaps you would be true, but that's the point, not everyone using the documentation will be as smart as you, or will start with your "distilled mindset on how to think about programs". Some people will be like me, will look at it, and say "that's tautological and doesn't really explain what's going on to me". After a few moments they may get it, but that's unnecessary friction.

Unless your documentation is purely for yourself, the mere presence of third parties that don't understand it means it is a failure. And again, you can argue until you're blue in the face that you prefer the documentation one way or another, but the fact that someone else didn't find it clear is not something you can _validly_ dismiss. Well, I suppose you could if you want to claim that the documentation is only intended for smart people and that no one else should use OCaml. If that's not your claim, you have to write documentation that other people understand, and merely because you find phrasing that they understand unpleasantly verbose because it uses two or three extra words makes no difference if they don't understand what you wrote.

Reply:
I like the "`f x` is `G x`" style for pure code; it encourages a functional mindset.  It's used quite a bit in the standard library already:

```
$ grep '(\*\*.*\] is \[' stdlib/*.mli
stdlib/bigarray.mli:  (** [size_in_bytes a] is [a]'s {!kind_size_in_bytes}. *)
…
stdlib/int32.mli:(** Successor.  [Int32.succ x] is [Int32.add x Int32.one]. *)
…
stdlib/listLabels.mli:(** [cons x xs] is [x :: xs]
…
stdlib/stdlib.mli:(** [succ x] is [x + 1]. *)
…
```

In these comments `is` tells you more than `returns` would; it says that the result of the function is all that there is to know about it.

However, the passive style is less helpful for imperative functions that have both a result and an effect, so I'd prefer not to see it here:

```ocaml
val iter : ('a -> unit) -> ('a, 'e) result -> unit
(** [iter f r] is [f v] if [r] is [Ok v] and [()] otherwise. *)
```

Perhaps `is` here could be changed to `calls` (e.g. `[iter f r] calls [f v] iff [r] is [Ok v]`).
```

Reply:
> The test here isn't whether you think the docs are okay. It's whether others find them confusing or not.

Neither is whether *you* think they are not; note that for now you seem to be the only one complaining. Over the years I have published almost 40'000 loc of commented `.mli` file using exactly this style and while I do get the occasional complaint about a confusing or unclear doc strings noone ever complained about this documentation style which is fairly natural to functional programmers.

Let me be clear now: I won't make any change to these doc strings so there's no need to continue this discussion especially since I don't think it helps up moving this PR forward.

Reply:
> Neither is whether you think they are not; note that for now you seem to be the only one complaining

I give up.